File: leetcode/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // buying (minPrice), selling (maxPrice) 
        int minPrice = Integer.MAX_VALUE; 
        int maxProfit = 0; 

        for(int i=0; i<prices.length; i++) {
            if(prices[i] < minPrice) {
                minPrice = prices[i];
            } else if(prices[i] - minPrice > maxProfit) {
                maxProfit = prices[i] - minPrice;
            }
        }
        
        return maxProfit;
    }
}

File: leetcode/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy = prices[0]
        profit = 0
        for i in range(1, len(prices)): 
            if prices[i] < buy:
                # update buy
                buy = prices[i]
            elif prices[i]-buy > profit: 
                # update profit if current profit is greater in this loop
                profit = prices[i]-buy
        return profit
